File: src/domain/normalize.ts
/**
 * الطبعنة (normalization) للبحث والمطابقة — spec/05.
 *
 * «طبعنة العربية واللاتينية للبحث **دون إتلاف النص الأصلي**.»
 * كل دوال هنا ترجع نسخة جديدة؛ النص الأصلي يُحفظ دائمًا في rawDescription.
 */

/** التشكيل والتطويل — تُزال للمطابقة فقط. */
const DIACRITICS = /[ؐ-ًؚ-ٰٟۖ-ۭـ]/g

/**
 * توحيد الحروف العربية المتغيّرة الشكل:
 * أإآا → ا · ى → ي · ة → ه · ؤ → و · ئ → ي
 * السبب: «عبد الفتاح» و«عبدالفتاح» و«عبد الفتّاح» اسم واحد في البحث.
 */
function unifyArabicLetters(text: string): string {
  return text
    .replace(/[أإآٱ]/g, 'ا')
    .replace(/ى/g, 'ي')
    .replace(/ة/g, 'ه')
    .replace(/ؤ/g, 'و')
    .replace(/ئ/g, 'ي')
}

/** الأرقام العربية والفارسية إلى لاتينية. */
export function latinizeDigits(text: string): string {
  let out = ''
  for (const ch of text) {
    const code = ch.codePointAt(0)!
    if (code >= 0x0660 && code <= 0x0669) out += String.fromCharCode(code - 0x0660 + 48)
    else if (code >= 0x06f0 && code <= 0x06f9) out += String.fromCharCode(code - 0x06f0 + 48)
    else out += ch
  }
  return out
}

/**
 * الصيغة المُطبعَنة المستخدمة في مطابقة القواعد والتجار والبحث.
 * تُحفظ إلى جانب النص الأصلي، ولا تحل محله.
 */
export function normalizeText(text: string): string {
  if (!text) return ''
  return unifyArabicLetters(latinizeDigits(text))
    .replace(DIACRITICS, '')
    .toUpperCase()
    .replace(/[^\p{L}\p{N}]+/gu, ' ') // كل ما ليس حرفًا أو رقمًا يصير مسافة
    .trim()
    .replace(/\s+/g, ' ')
}

/** صيغة مضغوطة بلا مسافات — تلتقط «عبدالفتاح» مقابل «عبد الفتاح». */
export function normalizeCompact(text: string): string {
  return normalizeText(text).replace(/\s/g, '')
}

/**
 * هل يحتوي النص المُطبعَن على العبارة المُطبعَنة؟
 * يقارن بالصيغتين العادية والمضغوطة حتى لا تفوت فروق المسافات.
 */
export function normalizedContains(haystack: string, needle: string): boolean {
  const n = normalizeText(needle)
  if (!n) return false
  if (normalizeText(haystack).includes(n)) return true
  return normalizeCompact(haystack).includes(normalizeCompact(needle))
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/Normalize.kt
package app.masroufy.core

/**
 * الطبعنة للبحث والمطابقة — نقل `src/domain/normalize.ts` (spec/05).
 * بترجّع نسخة جديدة دايمًا؛ النص الأصلي بيتحفظ زي ما هو.
 */

/** NFKC — أشكال العرض العربية (ﺔﻈﺣﻼﻣ من قارئ PDF قديم) ترجع حروفها. خاصة بكل جهاز. */
internal expect fun nfkc(text: String): String

/** التشكيل والتطويل — بيتشالوا للمطابقة بس. */
private fun isDiacritic(c: Char): Boolean =
    c.code in 0x0610..0x061A || c.code in 0x064B..0x065F || c.code == 0x0670 || c.code in 0x06D6..0x06ED || c.code == 0x0640

/** أإآٱ ⇒ ا · ى ⇒ ي · ة ⇒ ه · ؤ ⇒ و · ئ ⇒ ي («عبد الفتاح» و«عبدالفتّاح» اسم واحد في البحث). */
private fun unifyArabicLetter(c: Char): Char = when (c) {
    'أ', 'إ', 'آ', 'ٱ' -> 'ا'
    'ى' -> 'ي'
    'ة' -> 'ه'
    'ؤ' -> 'و'
    'ئ' -> 'ي'
    else -> c
}

/** الأرقام العربية والفارسية للاتينية. */
fun latinizeDigits(text: String): String {
    val out = StringBuilder(text.length)
    for (c in text) out.append(JsText.easternDigit(c) ?: c)
    return out.toString()
}

private val NOT_LETTER_OR_NUMBER = Regex("[^\\p{L}\\p{N}]+")
private val SPACES = Regex(" +")

/** الصيغة المطبعنة للمطابقة (القواعد والتجار والبحث) — بتتحفظ جنب الأصل مش مكانه. */
fun normalizeText(text: String): String {
    if (text.isEmpty()) return ""
    val unified = StringBuilder()
    for (c in latinizeDigits(nfkc(text))) {
        val u = unifyArabicLetter(c)
        if (!isDiacritic(u)) unified.append(u)
    }
    return unified.toString()
        .uppercase()
        .replace(NOT_LETTER_OR_NUMBER, " ")
        .trim()
        .replace(SPACES, " ")
}

/** من غير مسافات — «عبدالفتاح» قصاد «عبد الفتاح». */
fun normalizeCompact(text: String): String = normalizeText(text).replace(" ", "")

/** النص المطبعن فيه العبارة المطبعنة؟ بالصيغتين العادية والمضغوطة. */
fun normalizedContains(haystack: String, needle: String): Boolean {
    val n = normalizeText(needle)
    if (n.isEmpty()) return false
    if (normalizeText(haystack).contains(n)) return true
    return normalizeCompact(haystack).contains(normalizeCompact(needle))
}
